src/main/java/com/powsybl/caseserver/service/S3CaseService.java

```diff
@@ -0,0 +1,558 @@
+        return importer.getFormat();
+    }
+
+    // creates a directory, and then in this directory, initializes a file with content.
+    // After applying f to the file, deletes the file and the directory.
+    private <R, T1 extends Throwable, T2 extends Throwable> R withTempCopy(UUID caseUuid, String filename,
+                                                                                         FailableConsumer<Path, T1> contentInitializer, FailableFunction<Path, R, T2> f) {
+        Path tempdirPath;
+        Path tempCasePath;
+        try {
+            FileAttribute<Set<PosixFilePermission>> attr = PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString("rwx------"));
+            tempdirPath = Files.createTempDirectory(caseUuid.toString(), attr);
+
+            // Create parent directory if necessary
+            Path parentPath = Paths.get(filename).getParent();
+            if (parentPath != null) {
+                Files.createDirectory(tempdirPath.resolve(parentPath), attr);
+            }
+            // after this line, need to cleanup the dir
+        } catch (IOException e) {
+            throw CaseException.createTempDirectory(caseUuid, e);
+        }
+        try {
+            tempCasePath = tempdirPath.resolve(filename);
+            try {
+                contentInitializer.accept(tempCasePath);
+            } catch (CaseException e) {
+                throw CaseException.initTempFile(caseUuid, e);
+            } catch (Throwable ex) {
+                throw CaseException.initTempFile(caseUuid, ex);
+            }
+            // after this line, need to cleanup the file
+            try {
+                try {
+                    return f.apply(tempCasePath);
+                } catch (CaseException e) {
+                    throw CaseException.createFileNotImportable(tempdirPath);
+                } catch (Throwable t) {
+                    throw CaseException.processTempFile(caseUuid, t);
+                }
+            } finally {
+                try {
+                    Files.delete(tempCasePath);
+                } catch (IOException e) {
+                    LOGGER.error("Error cleaning up temporary case file", e);
+                }
+            }
+        } finally {
+            try {
+                if (Files.exists(tempdirPath)) {
+                    FileUtils.deleteDirectory(tempdirPath.toFile());
+                }
+            } catch (IOException e) {
+                LOGGER.error("Error cleaning up temporary case directory: " + tempdirPath, e);
+            }
+        }
+    }
+
+    // downloads from s3 and cleanup
+    public <R, T extends Throwable> R withS3DownloadedTempPath(UUID caseUuid, FailableFunction<Path, R, T> f) {
+        return withS3DownloadedTempPath(caseUuid, null, f);
+    }
+
+    public <R, T extends Throwable> R withS3DownloadedTempPath(UUID caseUuid, String caseFileKeyFromCaller, FailableFunction<Path, R, T> f) {
```

Comment:
why fromcaller ? all arguments are from callers, we usually don't put it in the variable name

Reply:
done